lib/Sema/LifetimeDependence.cpp

```diff
@@ -397,30 +433,16 @@ LifetimeDependenceInfo::infer(AbstractFunctionDecl *afd, Type resultType) {
   }
 
   if (afd->getKind() != DeclKind::Constructor && afd->hasImplicitSelfDecl()) {
-    ValueOwnership ownership = ValueOwnership::Default;
-    if (auto *AD = dyn_cast<AccessorDecl>(afd)) {
-      if (AD->getAccessorKind() == AccessorKind::Get ||
-          AD->getAccessorKind() == AccessorKind::Read) {
-        // We don't support "borrowing/consuming" ownership modifiers on
-        // getters/_read accessors, they are guaranteed by default for now.
-        ownership = ValueOwnership::Shared;
-      }
-      if (AD->getAccessorKind() == AccessorKind::Modify) {
-        ownership = ValueOwnership::InOut;
-      }
-    } else {
-      assert(afd->getKind() == DeclKind::Func);
-      ownership = afd->getImplicitSelfDecl()->getValueOwnership();
-      if (ownership == ValueOwnership::Default) {
-        diags.diagnose(
-            returnLoc,
-            diag::
-                lifetime_dependence_cannot_infer_wo_ownership_modifier_on_method);
-        return std::nullopt;
-      }
+    Type selfTypeInContext = dc->getSelfTypeInContext();
+    auto kind = getLifetimeDependenceKindFromType(selfTypeInContext);
+    auto selfOwnership = afd->getImplicitSelfDecl()->getValueOwnership();
+    if (!isLifetimeDependenceCompatibleWithOwnership(kind, selfOwnership,
+                                                     afd)) {
+      diags.diagnose(afd->getLoc(),
+                     diag::lifetime_dependence_invalid_self_ownership);
```

Comment:
What do you think about allowing lifetime dependence inference on implicit initializers when there is a single `~Escapable` type?

It's not very important, but could be nice, and probably more consistent with the proposal. Maybe in a separate follow-up PR.

Reply:
sure I'll do as a follow up, this can be done now that we got rid ownership modifier rules